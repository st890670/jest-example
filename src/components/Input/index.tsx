import React, { useCallback, useState, useMemo } from "react";

const Input = () => {
  const [inputValue, setInputValue] = useState<string>("");
  const [emptyWarning, setEmptyWarning] = useState<boolean>(false);

  const renderWarningText = useMemo(() => {
    if (!emptyWarning) {
      return <></>;
    }

    return (
      <div id="warningText" className="text-[#ff0000] text-sm mt-1 pl-3">
        必填
      </div>
    );
  }, [emptyWarning]);

  const handleInputChange = useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      const value = e.target.value;
      if (value) {
        setEmptyWarning(false);
      }
      setInputValue(value);
    },
    []
  );

  const handleKeyDown = useCallback(
    (e: React.KeyboardEvent) => {
      if (e.key !== "Enter") {
        return;
      }

      if (!inputValue) {
        setEmptyWarning(true);
        return;
      }
    },
    [inputValue]
  );

  return (
    <>
      <input
        className={`border rounded-[8px] py-1 px-3 focus:outline-none ${
          emptyWarning ? "border-[#ff0000]" : ""
        }`}
        onKeyDown={handleKeyDown}
        onChange={handleInputChange}
        placeholder="按Enter新增Todo"
      />
      {renderWarningText}
    </>
  );
};

export default Input;
