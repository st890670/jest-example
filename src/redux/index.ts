import { configureStore } from "@reduxjs/toolkit";
import createSagaMiddleware from "redux-saga";
import { combineReducers } from "redux";
import {
  useDispatch as useReduxDispatch,
  useSelector as useReduxSelector,
  TypedUseSelectorHook,
} from "react-redux";
import todoReduce from "./slice/todo";
import rootSaga from "./saga";

const sagaMiddleware = createSagaMiddleware();

const reducer = combineReducers({
    todo: todoReduce,
});

export const store = configureStore({
  reducer,
  middleware: [sagaMiddleware],
});

sagaMiddleware.run(rootSaga);

export const useDispatch = useReduxDispatch;
export const useSelector: TypedUseSelectorHook<ReturnType<typeof reducer>> =
  useReduxSelector;