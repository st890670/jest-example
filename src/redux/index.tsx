import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import createSagaMiddleware from "redux-saga";
import { combineReducers } from "redux";
import {
  useDispatch as useReduxDispatch,
  useSelector as useReduxSelector,
  TypedUseSelectorHook,
} from "react-redux";
import todoReduce from "./slices/todo";
import rootSaga from "./sagas";
import React from "react";

const sagaMiddleware = createSagaMiddleware();

export const reducer = combineReducers({
  todo: todoReduce,
});

const store = configureStore({
  reducer,
  middleware: [sagaMiddleware],
});

sagaMiddleware.run(rootSaga);

export const ReduxProvider = ({ children }: { children: React.ReactNode }) => (
  <Provider store={store}>{children}</Provider>
);

export const useDispatch = useReduxDispatch;
export const useSelector: TypedUseSelectorHook<ReturnType<typeof reducer>> =
  useReduxSelector;
