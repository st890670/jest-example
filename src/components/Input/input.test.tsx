import { shallow } from "enzyme";
import Input from "./index";
import * as redux from "../../redux";
import React from "react";

describe("Input Unit Test", () => {
  const useDispatchMock = jest.spyOn(redux, "useDispatch");
  beforeEach(() => {
    useDispatchMock.mockReturnValue(jest.fn());
  });

  it("Should match snapshot", () => {
    const component = shallow(<Input />);
    expect(component).toMatchSnapshot();
  });

  it("Should the input value same as the keydown value", () => {
    const component = shallow(<Input />);
    /**
     * 以下方程式碼為例:
     * 將find的結果DOM存成變數時要小心，如果針對這個變數進行模擬操作（simulate），
     * 必須重新find才會取得模擬後的值．
     *
     * const inputWrapper = component.find('#addInput')
     */
    const handleInputChange = component.find("#addInput").props()
      .onChange as React.ChangeEventHandler<HTMLInputElement>;
    handleInputChange({
      target: { value: "A" },
    } as React.ChangeEvent<HTMLInputElement>);
    expect(component.find("#addInput").getElement().props.value).toBe("A");
  });

  it("Should show warning when input value empty.", () => {
    const component = shallow(<Input />);
    const handleKeyDown = component.find("#addInput").props()
      .onKeyDown as React.KeyboardEventHandler<HTMLInputElement>;
    handleKeyDown({ key: "Enter" } as React.KeyboardEvent<HTMLInputElement>);
    expect(component.find("#warningText").hasClass("visible")).toBeTruthy();
  });

  it("Should hidden warning when input value existed.", () => {
    const component = shallow(<Input />);
    const handleKeyDown = component.find("#addInput").props()
      .onKeyDown as React.KeyboardEventHandler<HTMLInputElement>;
    handleKeyDown({ key: "A" } as React.KeyboardEvent<HTMLInputElement>);
    expect(component.find("#warningText").hasClass("invisible")).toBeTruthy();
  });
});

export {};
