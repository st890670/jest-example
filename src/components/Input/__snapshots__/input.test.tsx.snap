// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Unit Test Should match snapshot 1`] = `
<Fragment>
  <input
    className="border rounded-[8px] py-1 px-3 focus:outline-none "
    id="addInput"
    onChange={[Function]}
    onKeyDown={[Function]}
    placeholder="按Enter新增Todo"
    value=""
  />
  <div
    className="text-[#ff0000] text-sm mt-1 pl-1 invisible"
    id="warningText"
  >
    必填
  </div>
</Fragment>
`;
